File: 2025/9/9_25_leetcode_bc/104.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def findDepth(node, h):
            if not node:
                return h -1

            if not node.left and not node.right:
                return h


            left = findDepth(node.left, h+1)
            right = findDepth(node.right, h+1)

            return max(left, right)

        if not root:
            return 0
        return findDepth(root, 1)

File: 2025/9/9_25_leetcode_bc/104.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    const findDepth = (node, h) => {
        if (!node) return h - 1;

        if (!node.left && !node.right) return h;

        let left = findDepth(node.left, h+1);
        let right = findDepth(node.right, h+1);

        return Math.max(left, right);
    }

    if (!root) return 0;
    return findDepth(root, 1);
};